File: android/src/main/java/com/apollohg/editor/viewer/PreparedProseInstrumentation.kt
package com.apollohg.editor.viewer

import com.apollohg.editor.BuildConfig
import org.json.JSONArray
import org.json.JSONObject

/** Debug/device-only accounting, serialized at phase transitions and writes. */
internal object PreparedProseInstrumentation {
    enum class Owner { COMPILED, UNMOUNTED_LAYOUT, FABRIC_LEASE_HANDOFF, DIRECT_MOUNTED, IMAGE, SIDECARS, OTHER }
    enum class InvalidationReason { CONTENT, WIDTH, ATTACHMENT, FONT, MEMORY_PRESSURE, CACHE_RESET, REUSE }
    enum class TraversalPhase { COLD, WARM, IMAGES_DISABLED, RESET }
    private class PhaseSamples {
        val compileNanos = mutableListOf<Long>(); val layoutNanos = mutableListOf<Long>(); val combinedCompileLayoutNanos = mutableListOf<Long>()
        val cacheLookupNanos = mutableListOf<Long>(); val drawNanos = mutableListOf<Long>(); val frameNanos = mutableListOf<Long>(); val viewerFrameNanos = mutableListOf<Long>()
        var compileCount = 0; var layoutCount = 0; var cacheHits = 0; var cacheMisses = 0; var cacheWaits = 0; var drawCount = 0; var visibleBlocksDrawn = 0
        val invalidations = linkedMapOf<String, Int>()
        fun json() = JSONObject().put("compileNanos", JSONArray(compileNanos)).put("layoutNanos", JSONArray(layoutNanos)).put("combinedCompileLayoutNanos", JSONArray(combinedCompileLayoutNanos)).put("cacheLookupNanos", JSONArray(cacheLookupNanos)).put("drawNanos", JSONArray(drawNanos)).put("frameNanos", JSONArray(frameNanos)).put("viewerFrameNanos", JSONArray(viewerFrameNanos)).put("compileCount", compileCount).put("layoutCount", layoutCount).put("cacheHits", cacheHits).put("cacheMisses", cacheMisses).put("cacheWaits", cacheWaits).put("drawCount", drawCount).put("visibleBlocksDrawn", visibleBlocksDrawn).put("invalidations", JSONObject(invalidations as Map<*, *>))
    }
    private const val SAMPLE_LIMIT = 20_000; private val lock = Any()
    @Volatile private var enabled = false; private var phase: TraversalPhase? = null
    private val phaseSamples = linkedMapOf<TraversalPhase, PhaseSamples>(); private val completedPhases = linkedSetOf<TraversalPhase>(); private val pendingCompileNanos = linkedMapOf<TraversalPhase, MutableMap<String, Long>>()
    private val retained = linkedMapOf<String, Long>(); private var duplicatePublications = 0; private var previousFrameNanos = 0L; private var surfaceDrawnSinceFrame = false
    fun beginBenchmark() { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { enabled = true; resetLocked() } }
    fun reset() { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { resetLocked() } }
    fun beginPhase(next: TraversalPhase) { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { if (!enabled) return@synchronized; phase = next; completedPhases.remove(next); previousFrameNanos = 0; surfaceDrawnSinceFrame = false } }
    fun endPhase() { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { phase?.let(completedPhases::add); phase = null; previousFrameNanos = 0; surfaceDrawnSinceFrame = false } }
    /** Compatibility for device harnesses; bridge callers use begin/end phase. */
    fun beginTraversal(next: TraversalPhase) = beginPhase(next)
    fun endTraversal() = endPhase()
    fun onDisplayFrame(frameTimeNanos: Long) { if (!BuildConfig.PREPARED_PROSE_INSTRUMENTATION) return; synchronized(lock) { val active = phase ?: return; val previous = previousFrameNanos; previousFrameNanos = frameTimeNanos; if (previous == 0L || !surfaceDrawnSinceFrame) return; val elapsed = frameTimeNanos - previous; if (elapsed <= 0) return; val samples = samples(active); append(samples.frameNanos, elapsed); if (active == TraversalPhase.WARM) append(samples.viewerFrameNanos, elapsed); surfaceDrawnSinceFrame = false } }
    @JvmStatic fun exportJson(): String { if (!BuildConfig.PREPARED_PROSE_INSTRUMENTATION) return "{}"; return synchronized(lock) { val phases = JSONObject(); phaseSamples.filterKeys(completedPhases::contains).forEach { (phase, samples) -> phases.put(phase.name.lowercase(), samples.json()) }; val owners = Owner.values().associate { owner -> owner.name.lowercase() to retained.filterKeys { it.startsWith(owner.name + ":") }.values.sum() }; JSONObject().put("percentileDefinition", "nearest-rank: sorted[ceil(p*n)-1]").put("phaseSamples", phases).put("duplicatePublications", duplicatePublications).put("retainedBytes", JSONObject(owners)).toString() } }
    fun now(): Long = if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION && enabled) System.nanoTime() else 0L
    fun compiled(start: Long, generation: String) = record(start) { active, elapsed -> val samples = samples(active); samples.compileCount++; append(samples.compileNanos, elapsed); pendingCompileNanos.getOrPut(active) { linkedMapOf() }[generation] = elapsed }
    fun laidOut(start: Long, generation: String) = record(start) { active, elapsed -> val samples = samples(active); samples.layoutCount++; append(samples.layoutNanos, elapsed); pendingCompileNanos[active]?.remove(generation)?.let { append(samples.combinedCompileLayoutNanos, it + elapsed) } }
    fun cacheLookup(start: Long, hit: Boolean, waited: Boolean = false) = record(start) { active, elapsed -> val samples = samples(active); append(samples.cacheLookupNanos, elapsed); if (hit) samples.cacheHits++ else samples.cacheMisses++; if (waited) samples.cacheWaits++ }
    fun drew(start: Long, blocks: Int) = record(start) { active, elapsed -> val samples = samples(active); append(samples.drawNanos, elapsed); samples.drawCount++; samples.visibleBlocksDrawn += blocks; surfaceDrawnSinceFrame = true }
    fun retained(owner: Owner, scope: String, bytes: Long) { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION && enabled) synchronized(lock) { retained[owner.name + ":" + scope] = bytes.coerceAtLeast(0L) } }
    fun invalidated(reason: InvalidationReason) { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { if (enabled) phase?.let { active -> val values = samples(active).invalidations; values[reason.name] = values.getOrDefault(reason.name, 0) + 1 } } }
    fun duplicatePublication() { if (BuildConfig.PREPARED_PROSE_INSTRUMENTATION) synchronized(lock) { if (enabled) duplicatePublications++ } }
    private fun record(start: Long, block: (TraversalPhase, Long) -> Unit) { if (!BuildConfig.PREPARED_PROSE_INSTRUMENTATION || !enabled || start == 0L) return; synchronized(lock) { phase?.let { block(it, System.nanoTime() - start) } } }
    private fun samples(phase: TraversalPhase) = phaseSamples.getOrPut(phase) { PhaseSamples() }
    private fun append(target: MutableList<Long>, value: Long) { if (target.size < SAMPLE_LIMIT) target += value }
    private fun resetLocked() { phase = null; phaseSamples.clear(); completedPhases.clear(); pendingCompileNanos.clear(); retained.clear(); duplicatePublications = 0; previousFrameNanos = 0; surfaceDrawnSinceFrame = false }
}

File: ios/Viewer/PreparedProseInstrumentation.swift
import Foundation
import QuartzCore

/// Debug/device-only accounting. Samples belong to the traversal that caused
/// them; a phase transition is serialized with every producer under `lock`.
/// Percentiles use nearest-rank: sorted[ceil(p * n) - 1].
enum PreparedProseInstrumentation {
    enum Owner: String, CaseIterable { case compiled, unmountedLayout, fabricLeaseHandoff, directMounted, image, sidecars, other }
    enum InvalidationReason: String { case content, width, attachment, font, memoryPressure, cacheReset, reuse }
    enum TraversalPhase: String, CaseIterable { case cold, warm, imagesDisabled, reset }

    struct PhaseSamples: Codable {
        var compileNanos: [UInt64] = []; var layoutNanos: [UInt64] = []; var combinedCompileLayoutNanos: [UInt64] = []
        var cacheLookupNanos: [UInt64] = []; var drawNanos: [UInt64] = []; var frameNanos: [UInt64] = []; var viewerFrameNanos: [UInt64] = []
        var compileCount = 0; var layoutCount = 0; var cacheHits = 0; var cacheMisses = 0; var cacheWaits = 0; var drawCount = 0; var visibleBlocksDrawn = 0
        var invalidations: [String: Int] = [:]
    }
    struct Snapshot: Codable {
        let percentileDefinition: String; let phaseSamples: [String: PhaseSamples]
        let duplicatePublications: Int; let retainedBytes: [String: Int]
    }

    private static let lock = NSLock(); private static let sampleLimit = 20_000
    private static var enabled = false; private static var phase: TraversalPhase?
    private static var samples: [TraversalPhase: PhaseSamples] = [:]
    /// Exports must never observe a traversal while it is still collecting
    /// draw/frame evidence. A phase becomes visible only at `endPhase()`.
    private static var completedPhases: Set<TraversalPhase> = []
    /// Compile durations are associated by immutable generation, never by a
    /// positional zip of unrelated arrays.
    private static var pendingCompileNanos: [TraversalPhase: [String: UInt64]] = [:]
    private static var retainedByScope: [String: Int] = [:]; private static var duplicatePublications = 0
    private static var previousDisplayTimestamp: CFTimeInterval = 0; private static var surfaceDrawnSinceFrame = false

    static func beginBenchmark() {
#if DEBUG
        lock.lock(); enabled = true; resetLocked(); lock.unlock()
#endif
    }
    static func reset() {
#if DEBUG
        lock.lock(); resetLocked(); lock.unlock()
#endif
    }
    static func beginPhase(_ value: TraversalPhase) {
#if DEBUG
        lock.lock()
        guard enabled else { lock.unlock(); return }
        phase = value
        completedPhases.remove(value)
        previousDisplayTimestamp = 0
        surfaceDrawnSinceFrame = false
        lock.unlock()
#endif
    }
    static func endPhase() {
#if DEBUG
        lock.lock()
        if let phase { completedPhases.insert(phase) }
        phase = nil
        previousDisplayTimestamp = 0
        surfaceDrawnSinceFrame = false
        lock.unlock()
#endif
    }
    /// Native device harness compatibility; new bridges use the explicit
    /// begin/end phase names so cache reset cannot be mistaken for traversal.
    static func beginTraversal(_ value: TraversalPhase) { beginPhase(value) }
    static func endTraversal() { endPhase() }

    /// Called by CADisplayLink only while the collection traversal owns a
    /// mounted viewer surface. Frame evidence requires a preceding draw.
    static func displayLinkDidTick(_ displayLink: CADisplayLink) {
#if DEBUG
        lock.lock(); defer { lock.unlock() }
        guard let phase else { return }
        defer { previousDisplayTimestamp = displayLink.timestamp }
        guard previousDisplayTimestamp > 0, surfaceDrawnSinceFrame else { return }
        let nanos = UInt64((displayLink.timestamp - previousDisplayTimestamp) * 1_000_000_000)
        guard nanos > 0 else { return }
        mutate(phase) { samples in
            append(nanos, to: &samples.frameNanos)
            if phase == .warm { append(nanos, to: &samples.viewerFrameNanos) }
        }
        surfaceDrawnSinceFrame = false
#endif
    }

    static func exportJSON() -> String {
#if DEBUG
        lock.lock()
        let owners = Dictionary(uniqueKeysWithValues: Owner.allCases.map { owner in
            (owner.rawValue, retainedByScope.filter { $0.key.hasPrefix(owner.rawValue + ":") }.reduce(0) { $0 + $1.value })
        })
        let completedSamples = Dictionary(uniqueKeysWithValues: samples.compactMap { phase, samples in
            completedPhases.contains(phase) ? (phase.rawValue, samples) : nil
        })
        let snapshot = Snapshot(percentileDefinition: "nearest-rank: sorted[ceil(p*n)-1]", phaseSamples: completedSamples, duplicatePublications: duplicatePublications, retainedBytes: owners)
        lock.unlock()
        return String(data: (try? JSONEncoder().encode(snapshot)) ?? Data("{}".utf8), encoding: .utf8) ?? "{}"
#else
        return "{}"
#endif
    }

    @inline(__always) static func now() -> UInt64 {
#if DEBUG
        lock.lock(); let active = enabled; lock.unlock(); return active ? DispatchTime.now().uptimeNanoseconds : 0
#else
        return 0
#endif
    }
    @inline(__always) static func compiled(_ start: UInt64, generation: String) { record(start) { phase, elapsed in
        mutate(phase) { samples in samples.compileCount += 1; append(elapsed, to: &samples.compileNanos) }
        pendingCompileNanos[phase, default: [:]][generation] = elapsed
    } }
    @inline(__always) static func laidOut(_ start: UInt64, generation: String) { record(start) { phase, elapsed in
        mutate(phase) { samples in
            samples.layoutCount += 1; append(elapsed, to: &samples.layoutNanos)
            if let compile = pendingCompileNanos[phase]?.removeValue(forKey: generation) { append(compile + elapsed, to: &samples.combinedCompileLayoutNanos) }
        }
    } }
    @inline(__always) static func cacheLookup(_ start: UInt64, hit: Bool, waited: Bool = false) { record(start) { phase, elapsed in mutate(phase) { samples in append(elapsed, to: &samples.cacheLookupNanos); if hit { samples.cacheHits += 1 } else { samples.cacheMisses += 1 }; if waited { samples.cacheWaits += 1 } } } }
    @inline(__always) static func drew(_ start: UInt64, visibleBlocks: Int) { record(start) { phase, elapsed in mutate(phase) { samples in append(elapsed, to: &samples.drawNanos); samples.drawCount += 1; samples.visibleBlocksDrawn += visibleBlocks }; surfaceDrawnSinceFrame = true } }
    static func retained(_ owner: Owner, scope: String, bytes: Int) {
#if DEBUG
        lock.lock(); if enabled { retainedByScope[owner.rawValue + ":" + scope] = max(0, bytes) }; lock.unlock()
#endif
    }
    static func invalidated(_ reason: InvalidationReason) {
#if DEBUG
        lock.lock(); if enabled, let phase { mutate(phase) { samples in samples.invalidations[reason.rawValue, default: 0] += 1 } }; lock.unlock()
#endif
    }
    static func duplicatePublication() {
#if DEBUG
        lock.lock(); if enabled { duplicatePublications += 1 }; lock.unlock()
#endif
    }
    private static func record(_ start: UInt64, _ body: (TraversalPhase, UInt64) -> Void) {
#if DEBUG
        guard start != 0 else { return }; lock.lock(); defer { lock.unlock() }; guard enabled, let phase else { return }; body(phase, DispatchTime.now().uptimeNanoseconds - start)
#endif
    }
    private static func mutate(_ phase: TraversalPhase, _ body: (inout PhaseSamples) -> Void) { var value = samples[phase] ?? PhaseSamples(); body(&value); samples[phase] = value }
    private static func append(_ value: UInt64, to samples: inout [UInt64]) { if samples.count < sampleLimit { samples.append(value) } }
    private static func resetLocked() { phase = nil; samples = [:]; completedPhases = []; pendingCompileNanos = [:]; retainedByScope = [:]; duplicatePublications = 0; previousDisplayTimestamp = 0; surfaceDrawnSinceFrame = false }
}
